File: source/sort/merge/java/MergeSort.java
class MergeSort {
    public static void merge(int[] arr, int left, int mid, int right) {
        int llen = mid - left + 1;
        int rlen = right - mid;
        int larr[] = new int [llen];
        int rarr[] = new int [rlen];
        for (int i = 0; i < llen; ++i)
            larr[i] = arr[left + i];
        for (int j = 0; j < rlen; ++j)
            rarr[j] = arr[mid + 1 + j];
        int k = left;
        int i = 0;
        int j = 0;
        while (i < llen && j < rlen) {
            if (larr[i] <= rarr[j]) {
                arr[k] = larr[i];
                i++;
            } else {
                arr[k] = rarr[j];
                j++;
            }
            k++;
        }
        while (i < llen) {
            arr[k] = larr[i];
            i++;
            k++;
        }
        while (j < rlen) {
            arr[k] = rarr[j];
            j++;
            k++;
        }
    }
    public static void mergeSort(int[] arr, int left, int right) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergeSort(arr, left, mid);
            mergeSort(arr , mid + 1, right);
            merge(arr, left, mid, right);
        }
    }
}

File: source/sort/merge/py/mergesort.py
def merge(arr, left, mid, right):
    llen = mid - left + 1
    rlen = right - mid
    larr = [0] * (llen)
    rarr = [0] * (rlen)
    for i in range(0, llen):
        larr[i] = arr[left + i]
    for j in range(0, rlen):
        rarr[j] = arr[mid + 1 + j]
    i = 0
    j = 0
    k = left
    while i < llen and j < rlen:
        if larr[i] <= rarr[j]:
            arr[k] = larr[i]
            i += 1
        else:
            arr[k] = rarr[j]
            j += 1
        k += 1
    while i < llen:
        arr[k] = larr[i]
        i += 1
        k += 1
    while j < rlen:
        arr[k] = rarr[j]
        j += 1
        k += 1


def merge_sort(arr, left, right):
    if left < right:
        mid = (left + (right - 1)) / 2
        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)
        merge(arr, left, mid, right)
